UnitystationLauncher/ContentScanning/AssemblyTypeChecker.cs

```diff
@@ -0,0 +1,1211 @@
+                                {
+                                    var a = mMemberRefMethod.ParameterTypes[i];
+                                    var b = parsed.ParameterTypes[i];
+
+                                    if (!a.WhitelistEquals(b))
+                                    {
+                                        notParamMismatch = false;
+                                        break;
+
+                                    }
+                                }
+
+                                if (notParamMismatch)
+                                {
+                                    break; // Found
+                                }
+                                break;
+                            }
+                        }
+
+                        if (notParamMismatch == false)
+                        {
+                            continue;
+                        }
+
+                        errors.Add(new SandboxError($"Access to method not allowed: {mMemberRefMethod}"));
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(memberRef));
+                }
+            }
+        }
+    }
+
+    private void CheckInheritance(
+        SandboxConfig sandboxConfig,
+        List<(MType type, MType parent, ArraySegment<MType> interfaceImpls)> inherited,
+        ConcurrentBag<SandboxError> errors)
+    {
+        // This inheritance whitelisting primarily serves to avoid content doing funny stuff
+        // by e.g. inheriting Type.
+        foreach (var (_, baseType, interfaces) in inherited)
+        {
+            if (CanInherit(baseType) == false)
+            {
+                errors.Add(new SandboxError($"Inheriting of type not allowed: {baseType}"));
+            }
+
+            foreach (var @interface in interfaces)
+            {
+                if (CanInherit(@interface) == false)
+                {
+                    errors.Add(new SandboxError($"Implementing of interface not allowed: {@interface}"));
+                }
+            }
+
+            bool CanInherit(MType inheritType)
+            {
+                var realBaseType = inheritType switch
+                {
+                    ScanningTypes.MTypeGeneric generic => (ScanningTypes.MTypeReferenced)generic.GenericType,
+                    ScanningTypes.MTypeReferenced referenced => referenced,
+                    _ => throw new InvalidOperationException() // Can't happen.
+                };
+
+                if (IsTypeAccessAllowed(sandboxConfig, realBaseType, out var cfg) == false)
+                {
+                    return false;
+                }
+
+                return cfg.Inherit != InheritMode.Block && (cfg.Inherit == InheritMode.Allow || cfg.All);
+            }
+        }
+    }
+
+    private bool IsTypeAccessAllowed(SandboxConfig sandboxConfig, ScanningTypes.MTypeReferenced type,
+        [NotNullWhen(true)] out TypeConfig? cfg)
```

Comment:
Is this `[NotNullWhen(true)]` just to get rid of a warning? I'm not sure why its here

Reply:
looks like it's inherited from SS14 code